torchao/quantization/prototype/mixed_precision/scripts/BO_acc.py

```diff
@@ -0,0 +1,345 @@
+            raw_data=eval_results, 
+        )
+
+        del m
+        torch.cuda.empty_cache()
+
+
+    print("------Finish BO------")
+    for h in history:
+        print(h)
+    write_history_to_csv(history, output_file, ["cal_PPL", "model_size", "quant_config"])
+
+    print("------Best config------")
+    best_parameters, values = ax_client.get_best_parameters()
+    print(best_parameters, values)
+
+# Worker function to perform BO trials on a specific GPU
+def eval_in_parallel(gpu_id, checkpoint, num_PPL_eval_samples, config, return_dict, proc_id, trial_id):
+
+    model, tokenizer = load_model(checkpoint, f'cuda:{gpu_id}')
+    parameters_list = define_parameter_list()
+
+    print(f"Process {proc_id} on GPU {gpu_id} starts!")
+
+    quantize_by_fqn_to_config(model=model, device=f'cuda:{gpu_id}', fqn_to_config=dict(config))
+
+    eval_results = eval(model, tokenizer, num_PPL_eval_samples, config)
+        
+    return_dict[proc_id] = (trial_id, config, eval_results)
+    
+    del model
+    torch.cuda.empty_cache()
+
+'''
+This function will run BO trials in parallel on multiple GPUs.
+Each time the BO gets multiple new trials, evaluates the trials on the GPUs and return the evaluation results to update the BO.
+Multiple trials, one BO update.
+'''
+def run_parallel_BO(device, checkpoint, num_PPL_eval_samples, num_BO_initial_samples, num_trials, model_size_constraint, gpu_list):
+
+    parameters_list = define_parameter_list()
+    initial_points_set = get_initial_samples(num_BO_initial_samples)
+
+    #initialize ax_client
+    constraint="model_size <= "+str(model_size_constraint)
+    ax_client = AxClient()
+    ax_client.create_experiment(
+        parameters = parameters_list,
+        name = "test_quantize_BO",
+        objectives = {"cal_PPL": ObjectiveProperties(minimize=True)},
+        choose_generation_strategy_kwargs = {
+            "num_BO_initial_samplesization_trials": num_BO_initial_samples, # the number of trials to build generation strategy
+        },
+        outcome_constraints=[constraint],
+    )
+
+    gpu_list = [int(i) for i in gpu_list.split(",")]
+
+    history=[]
+    trial_id = 0
+
+    # Set the multiprocessing start method to 'spawn'
+    mp.set_start_method("spawn", force=True)
+
+    # add initial points into the BO trials
+    for id in range(num_BO_initial_samples//len(gpu_list)):
+        processes = []
+        manager = mp.Manager()
+        return_dict = manager.dict()
+
+        # Start the worker processes
+        for i, gpu_id in enumerate(gpu_list):
+            ax_client.attach_trial(parameters=dict(initial_points_set[id*len(gpu_list)+i]))
+            p = mp.Process(target=eval_in_parallel, args=(gpu_id, checkpoint, num_PPL_eval_samples, initial_points_set[id*len(gpu_list)+i], return_dict, i, trial_id))
+            trial_id += 1
+            p.start()
+            processes.append(p)
+
+        # Wait for all processes to finish
+        for p in processes:
+            p.join()
```

Comment:
outside the scope of this PR, but in the future we may want to consider a pattern here that completes a trial + generates a new config + launches a new trial whenever one of the running processes is complete. This is particularly important if there is large heterogeneity in the evaluation times between trials.